tensilelite/rocisa/rocisa/include/container.hpp

```diff
@@ -0,0 +1,1086 @@
+/* ************************************************************************
+ * Copyright (C) 2025 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * ************************************************************************ */
+#pragma once
+#include <algorithm>
+#include <cassert>
+#include <cmath>
+#include <memory>
+#include <numeric>
+#include <optional>
+#include <sstream>
+#include <string>
+#include <vector>
+
+#include "base.hpp"
+#include "enum.hpp"
+
+namespace rocisa
+{
+    struct Container
+    {
+        Container() {}
+
+        virtual ~Container() = default;
+
+        virtual std::shared_ptr<Container> clone() const = 0;
+
+        virtual std::string toString() const = 0;
+    };
+
+    struct DSModifiers : public Container
+    {
+        DSModifiers(int na = 1, int offset = 0, int offset0 = 0, int offset1 = 0, bool gds = false)
+            : Container()
+            , na(na)
+            , offset(offset)
+            , offset0(offset0)
+            , offset1(offset1)
+            , gds(gds)
```

Comment:
Just a comment. We probably don't need the `gds` field (not sure if its actually [supported](https://github.com/ROCm/ROCm/issues/1296#issuecomment-1338772042)..)